### DISCUSSION TITLE
Run DataMigration without creating module

### DISCUSSION THREAD

**Discussion by EricBui619** [OrchardCMS/OrchardCore#9154] [Q&A] [answered]
I created an Orchard Core project with a data migration to create some content types with its content parts and filed. So I want migration to be executed for first time project starting but I got an error, IContentDefinitionManager can not be resolved. When creating a separated module with enabled Content module, I didn't get any problem. How can I solve it? How about recipe? Can I use it for enabling some module first time?

**Comment by jtkech** [comment, marked as answer]
@EricArita 

So i assume you defined a `Migrations` from your main app and registered like this

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddOrchardCms()
                .ConfigureServices(s =>
                {
                    s.AddScoped<IDataMigration, Migrations>();
                });
        }

I tried it, works well but not on a fresh starup, the problem is that the application behaves as a module but that is always enabled, so the migrations will be activated during the setup as soon as we create a first shell scope and before any recipe execution and other features enabling including the one that register the content definition manager.

I will check it this night and will let you know ;)

**Reply by jtkech** [reply]
@EricArita 

So as said the migrations defined from the application is resolved too early, before the content definition manager is registered by the `OC.Contents` module in a subsequent scope. If defined through a module feature, usualy we define a dependency for this feature (can be done through the application but keep it simple), or we can use the `[RequireFeatures("OrchardCore.Contents")]` on a specific startup for this feature, the one that registers the migrations.

But here we can't use the `[RequireFeatures("OrchardCore.Contents")]` as we use the `ConfigureServices()` helper, so here the workaround is to resolve lazily the content definition manager and check if it is null.

    public class Migrations : DataMigration
    {
        private IServiceProvider _serviceProvider;
        private IContentDefinitionManager _contentDefinitionManager;

        public Migrations(IServiceProvider serviceProvider)// IContentDefinitionManager contentDefinitionManager)
        {
            //_contentDefinitionManager = contentDefinitionManager;
            _serviceProvider = serviceProvider;
        }

        public int Create()
        {
            _contentDefinitionManager ??= _serviceProvider.GetService<IContentDefinitionManager>();
            if (_contentDefinitionManager == null)
            {
                return 0;
            }

            _contentDefinitionManager.AlterPartDefinition("CommonPart", builder => builder
                .Attachable()
                .WithDescription("Testing: Provides an editor for the common properties of a content item."));

            return 1;
        }
    }

Hmm, but i will also check if we can define a custom startup from the application and with the `RequireFeatures()`, i will let you know.

**Reply by jtkech** [reply]
Okay it works, beautiful ;)

From your application you just need to also define

    [RequireFeatures("OrchardCore.Contents")]
    public class MigrationsStartup : StartupBase
    {
        // You can adjust the order if you want that your migrations runs later.
        public override int Order => 150;

        public override void ConfigureServices(IServiceCollection services)
        {
            services.AddScoped<IDataMigration, Migrations>();
        }
    }

**Reply by EricBui619** [reply]
@jtkech Oh Thank you for your support ! I will give it a shot and let you know

**Reply by EricBui619** [reply]
@jtkech hmm I've tried the [RequireFeatures("OrchardCore.Contents")] attribute but it seems not to work in the Startup class of the main app as I want, I got the error with content definition manager again. Like you did above, It looks like a Startup for a module, right? I think with the one we can solve better by using Manifest.cs file instead.
After considering, I decided to choose  to create my custom module and enable it to trigger migration for keeping everything simple. Btw, I wonder how can I enable my custom module as a feature as soon as project starts up?

**Reply by jtkech** [reply]
> @jtkech hmm I've tried the [RequireFeatures("OrchardCore.Contents")] attribute but it seems not to work in the Startup class of the main app as I want

No, don't decorate your main startup of your app with this require attribute, just define in any file of your app a standalone custom startup class that will be detected by OC, may be in the same file where you defined your migrations

And then don't register your migrations through the main startup class anymore, but through this dedicated startup

I tried it, it works

**Reply by EricBui619** [reply]
Ya I got it, thank you so much :D
